Lib/json/__init__.py

```diff
@@ -242,12 +242,13 @@ def dumps(obj, *, skipkeys=False, ensure_ascii=True, check_circular=True,
 
 
 def detect_encoding(b):
-    bstartswith = b.startswith
-    if bstartswith((codecs.BOM_UTF32_BE, codecs.BOM_UTF32_LE)):
-        return 'utf-32'
-    if bstartswith((codecs.BOM_UTF16_BE, codecs.BOM_UTF16_LE)):
-        return 'utf-16'
-    if bstartswith(codecs.BOM_UTF8):
+    for prefix in (codecs.BOM_UTF32_BE, codecs.BOM_UTF32_LE):
+        if b[:len(prefix)] == prefix:
```

Comment:
purpose of this change is to allow to use c-ext classes that supports buffer protocol.

I can remove memoryview from allowed type, and if encoding detection fails then fail with TypeError.

proper solution would be to write memoryview like object that supports only "b|B" format